alpaka does not work with Clang 17+ as CUDA compiler with vanila libstdc++ (gnu)

CMake configure fails if I use Clang 17+ as CUDA compiler together with the `libstdc++` library (part of the GCC compiler collection).

```bash
$ cmake --preset gpu-cuda-clang 
Preset CMake variables:

  BUILD_TESTING="ON"
  CMAKE_CUDA_COMPILER="clang++"
  CMAKE_CXX_COMPILER="clang++"
  alpaka_ACC_CPU_B_SEQ_T_SEQ_ENABLE="ON"
  alpaka_ACC_GPU_CUDA_ENABLE="ON"
  alpaka_BUILD_EXAMPLES="ON"

-- The CXX compiler identification is Clang 17.0.4
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/spack-modules/opt/spack/linux-ubuntu22.04-zen2/gcc-11.3.0/llvm-17.0.4-fdedxkcbid2axx2csrbi3bsjobdhcdky/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found Boost: /opt/spack-modules/opt/spack/linux-ubuntu22.04-zen2/gcc-11.3.0/boost-1.75.0-z3egc22qy3an5l2p6ojcdg4cc6iqvvxe/lib/cmake/Boost-1.75.0/BoostConfig.cmake (found suitable version "1.75.0", minimum required is "1.74.0") found components: atomic 
-- boost::atomic_ref<T> found
-- C++20 math constants not found. Falling back to non-standard constants.
-- The CUDA compiler identification is unknown
CMake Error at /opt/spack-modules/opt/spack/linux-ubuntu22.04-zen2/gcc-12.2.0/cmake-3.26.6-b3nwcvfcgtjtzufrlhbl2ost5ybbfk2n/share/cmake-3.26/Modules/CMakeDetermineCUDACompiler.cmake:603 (message):
  Failed to detect a default CUDA architecture.



  Compiler output:

Call Stack (most recent call first):
  cmake/alpakaCommon.cmake:376 (enable_language)
  CMakeLists.txt:95 (include)


-- Configuring incomplete, errors occurred!
```

The reason is because CUDA does not support float 128bit, which is part of the GNU extensions of the  `libstdc++` . Details are described in this issue: https://gitlab.kitware.com/cmake/cmake/-/issues/25861

Since Clang 17, the default C++ standard is `gnu++17`. Before it was `c++14`. https://reviews.llvm.org/D155539

The bug does not appears, if you use the `libstdc++` installed via `apt` on Debian based operation systems, because they apply a patch to solve the problem: 
* https://salsa.debian.org/toolchain-team/gcc/-/blob/32556e0c070cdd077a1289caa85d990080cd43c0/debian/patches/cuda-float128.diff
* https://bugs.launchpad.net/ubuntu/+source/glibc/+bug/1717257

Therefore the problem is not detected in our CI, because we use Ubuntu based container and install `gcc` via `apt`.

The problem can be reproduced on the `HAL` workstation, because it use a `libstdc++` version installed via spack.

--- a/cmake/alpakaCommon.cmake
+++ b/cmake/alpakaCommon.cmake
@@ -373,7 +373,44 @@ if(alpaka_ACC_GPU_CUDA_ENABLE)
             endif()
         endif()
 
+        # the CMake compiler detection of clang 17 and 18 as CUDA compiler is broken
+        # the detection try to compile an empty file with the default C++ standard of clang, which is -std=gnu++17
+        # but CUDA does not support the 128 bit float extension, therefore the test failes
+        # more details: https://gitlab.kitware.com/cmake/cmake/-/issues/25861
+        # this workaround disable the gnu extensions for the compiler detection
+        # the bug is fixed in clang 19: https://github.com/llvm/llvm-project/issues/88695
+        if("${CMAKE_CUDA_COMPILER}" MATCHES "clang*")
+            # get compiler version without enable_language()
+            execute_process(COMMAND ${CMAKE_CUDA_COMPILER} -dumpversion
+                   OUTPUT_VARIABLE _CLANG_CUDA_VERSION
+                   RESULT_VARIABLE _CLANG_CUDA_VERSION_ERROR_CODE)
+
+            if(NOT "${_CLANG_CUDA_VERSION_ERROR_CODE}" STREQUAL "0")
+                message(FATAL_ERROR "running '${CMAKE_CUDA_COMPILER} -dumpversion' failed: ${_CLANG_CUDA_VERSION_ERROR_CODE}")
+            endif()
+
+            string(STRIP ${_CLANG_CUDA_VERSION} _CLANG_CUDA_VERSION)
+            message(DEBUG "Workaround: manual checked Clang-CUDA version: ${_CLANG_CUDA_VERSION}")
+
+            if(${_CLANG_CUDA_VERSION} VERSION_GREATER_EQUAL 17 AND ${_CLANG_CUDA_VERSION} VERSION_LESS 19)
+                message(DEBUG "Workaround: apply -std=c++98 for clang as cuda compiler")
+                set(_CMAKE_CUDA_FLAGS_BEFORE ${CMAKE_CUDA_FLAGS})
+                # we need to use C++ 98 for the detection test, because from new, disabling the extension is ignored for C++ 98
+                set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -std=c++98")
+            endif()
+        endif()
+
         enable_language(CUDA)
+
+        if(DEFINED _CLANG_CUDA_VERSION)
+            message(DEBUG "Workaround: reset variables for clang as cuda compiler -std=c++98 fix")
+            # remove the flag compiler -std=c++98
+            set(CMAKE_CUDA_FLAGS ${_CMAKE_CUDA_FLAGS_BEFORE})
+            unset(_CMAKE_CUDA_FLAGS_BEFORE)
+            unset(_CLANG_CUDA_VERSION)
+            unset(_CLANG_CUDA_VERSION_ERROR_CODE)
+        endif()
+
         find_package(CUDAToolkit REQUIRED)
 
         target_compile_features(alpaka INTERFACE cuda_std_${alpaka_CXX_STANDARD})

